src/TicketingSystem.PurchasesAPI/Controllers/OrdersController.cs

```diff
@@ -0,0 +1,186 @@
+        [Route("carts/{cartId}/events/{eventId}/seats/{seatId}")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        public async Task<IActionResult> AddSeatToCart([FromRoute] string cartId, [FromRoute] string eventId, [FromRoute] string seatId)
+        {
+            var existingEvent = await _eventService.GetByIdAsync(eventId);
+            if(existingEvent is null)
+            {
+                return NotFound(nameof(eventId));
+            }
+
+            var existingSeatId = await _eventSeatService.GetByIdAsync(seatId);
+            if (existingSeatId is null)
+            {
+                return NotFound(nameof(seatId));
+            }
+
+            return await DeleteSeatFromCart(seatId, cartId);
+        }
+
+        /// <summary>
+        /// Moves all the seats in the cart to a booked state. Returns the ID of the related Payment
+        /// </summary>
+        [HttpPut]
+        [Route("carts/{cartId}/book")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        public async Task<IActionResult> BookSeatsInCart([FromRoute] string cartId)
+        {
+            var seats = (await _cartItemService.FilterAsync(s => s.CartId == cartId))?.Select(ci => ci.EventSeatId).ToList();
+
+            await _eventSeatService.UpdateEventSeatsStates(seats ?? Enumerable.Empty<string>().ToList(), EventSeatState.Booked);
+
+            var paymentsOfCart = await _paymentService.FilterAsync(p => p.CartId == cartId);
+            var payment = paymentsOfCart?.FirstOrDefault();
+
+            if (paymentsOfCart == null || payment == null)
+            {
+                return NotFound(nameof(paymentsOfCart)); // TODO Verify that payment is created during other transactions
+            }
+
+            return Ok(payment.Id);
+        }
+
+        private async Task<SeatStateModel> AddSeatToCart(AddCartModel model, string cartId)
+        {
+            var ticket = new TicketDto
+            {
+                EventSeatId = model.SeatId,
+                EventId = model.EventId,
+                State = BusinessLogic.Enums.TicketState.NotPurchased,
+                PriceOption = model.PriceOption,
+                Price = model.Price,
+                PurchasedOn = null,
+                UserId = model.UserId,
+            };
+
+            await _ticketService.CreateAsync(ticket);
+
+            var cartItem = new CartItemDto
+            {
+                CartId = cartId,
+                TicketId = ticket.Id,
+                EventSeatId = model.SeatId,
+                CreatedOn = DateTimeOffset.UtcNow
+            };
+
+            await _cartItemService.CreateAsync(cartItem);
+
+            var cartItems = await _cartItemService.FilterAsync(ci => ci.CartId == cartId);
+
+            var payment = await _paymentService.UpsertInProgressPayment(cartId, cartItems.Select(x => x.Id).ToArray());
+
+            return new SeatStateModel
+            {
```

Comment:
The same suggestion about logic in controllers - later on we will have additional checks for seat state, payment state, some transaction logic, ... I think it is better having it in Services in the BLL

Reply:
Moved extra logic from controllers to services